sdk/core/src/main/java/org/apache/mesos/config/ConfigurationUpdater.java

```diff
@@ -0,0 +1,230 @@
+        final List<ConfigurationValidationError> errors = new ArrayList<>();
+        if (targetConfig != null) {
+            try {
+                LOGGER.info("Old target config: {}", targetConfig.toJsonString());
+            } catch (Exception e) {
+                LOGGER.error("Unable to get JSON representation of old target config object", e);
+                // Don't add a validation error: That'd prevent the new config from replacing this
+                // one, and we'd be stuck with this config forever! Hopefully the new config will fix things...
+            }
+        } else {
+            LOGGER.info("Old target config: <null>");
+        }
+        try {
+            LOGGER.info("New prospective config: {}", candidateConfig.toJsonString());
+        } catch (Exception e) {
+            LOGGER.error("Unable to get JSON representation of new prospective config object: " + candidateConfig, e);
+            errors.add(ConfigurationValidationError.valueError("NewConfigAsJson", "jsonString",
+                    String.format("Unable to serialize new config to JSON for logging: %s", e.getMessage())));
+        }
+
+        // Check for any validation errors (including against the prior config, if one is available)
+        // NOTE: We ALWAYS run validation regardless of config equality. This allows the developer
+        // to e.g. require that a value be increased, whereas a short-circuit on equality would
+        // prevent that sort of validation.
+        for (ConfigurationValidator<C> validator : validators) {
+            errors.addAll(validator.validate(targetConfig, candidateConfig));
+        }
+
+        // Select the appropriate configuration ID as the target. If the config hasn't changed or if
+        // there are validation errors against the new config, we continue using the prior target.
+        if (!errors.isEmpty()) {
+            StringJoiner sj = new StringJoiner("\n");
+            int i = 1;
+            for (ConfigurationValidationError error : errors) {
+                sj.add(String.format("%d: %s", i++, error.toString()));
+            }
+            LOGGER.warn("New configuration failed validation against current target " +
+                    "configuration {}, with {} errors across {} validators:\n{}",
+                    targetConfigId, errors.size(), validators.size(), sj.toString());
+            if (targetConfig == null) {
+                throw new ConfigStoreException(String.format(
+                        "Configuration failed validation without any prior target configuration" +
+                        "available for fallback. Initial launch with invalid configuration? " +
+                        "%d Errors: %s", errors.size(), sj.toString()));
+            }
+        } else if (targetConfig == null || !configComparer.equals(targetConfig, candidateConfig)) {
+            LOGGER.info("Changes detected between current target configuration '{}' and new " +
```

Comment:
Old and new configs are both outputted in JSON form earlier in the file. (lines 98-115)